### PR TITLE
Add source file location to jazzer stack traces

### PR DESCRIPTION
**Pull request by florianGla**
This PR uses a source map to convert the location of jazzer findings into file paths.

Consider our maven example project.
Previous finding location: com.example.ExploreMe:19
New finding location: src/main/java/com/example/ExploreMe.java:19

### Motivation/Context

This is needed to be able to map a finding to a location in the source code.

### How to use/reproduce

Start a fuzzing run for a java project.

### CODE REVIEW

File: pkg/parser/libfuzzer/stacktrace/stacktrace.go [hunk 1]
```diff
@@ -265,3 +294,28 @@ func (p *parser) sourceLocationFromLine(line string) (*StackFrame, error) {
...
 }
+
+func (p *parser) setJavaSourceFilePath(frame *StackFrame) {
+	removeLastPart := func(javaName string) string {
+		sepIndex := strings.LastIndex(javaName, ".")
+		if sepIndex > 0 {
+			return javaName[0:sepIndex]
+		}
+		return ""
+	}
```
Comment (jochil) [inline pkg/parser/libfuzzer/stacktrace/stacktrace.go:308]: I would prefer moving this to a regular function and write a short test for it

**Review by jochil — COMMENTED** [review]
I have the feeling that this change can use some more (unit) tests... there are a lot of changes to the code but little new tests

In general the solution with the `SourceFiles` field in the parser makes sense for me

File: internal/build/java/java.go [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+			return nil, err
+		}
+		if sourceDir != "" {
+			sourceDirs = append(sourceDirs, sourceDir)
+		}
+	} else {
+		sourceDirs = append(sourceDirs, filepath.Join(projectDir, "src", "main"))
+	}
+
+	return sourceDirs, nil
```
[the remark this answers is not in the document]
Comment (florianGla) [reply internal/build/java/java.go:31]: Yes, `sourceDirs` are only returned at the end of the function.

File: pkg/parser/libfuzzer/stacktrace/stacktrace.go [hunk 3]
```diff
@@ -164,13 +186,21 @@ func (p *parser) stackFrameFromLine(line string) (*StackFrame, error) {
...
 		SourceFile:  filepath.ToSlash(sourceFile),
 		Line:        uint32(lineNumber),
 		Column:      uint32(column),
 		FrameNumber: uint32(frameNumber),
-		Function:    matches["function"],
-	}, nil
+		Function:    function,
+	}
+
+	log.Printf("xxxxxxxxxxxx %v", stackFrame)
```
Comment (jochil) [inline pkg/parser/libfuzzer/stacktrace/stacktrace.go:200]: I guess this is a left over

File: pkg/java/source_map.go [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+	"github.com/pkg/errors"
+)
+
+// SourceMap provides a mapping from package names
+// into the corresponding source file locations
+type SourceMap struct {
+	JavaPackages map[string][]string `json:"java_packages,omitempty"`
+}
+
+func CreateSourceMap(projectDir string, sourceDirs []string) (*SourceMap, error) {
```
Comment (jochil) [inline pkg/java/source_map.go:18]: I think we should cover this function by a test too... there is a lot of stuff happening here

File: pkg/java/source_map.go [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+	}
+	for _, file := range sourceFiles {
+		packageName, err := func() (string, error) {
+			fd, err := os.Open(file)
+			if err != nil {
+				return "", errors.WithStack(err)
+			}
+			defer fd.Close()
+			return GetPackageFromSource(fd), nil
+		}()
```
Comment (jochil) [inline pkg/java/source_map.go:40]: you are basically re-declaring this function over and over in a loop... why not extracting it and using `file` as a parameter?

File: pkg/java/source_map.go (hunk 4 above)
Comment (florianGla) [reply pkg/java/source_map.go:18]: I can do that. I hoped that it is already covered by the integration tests for maven/gradle and jazzer_bundler. We would need an integration test with real test data.

File: pkg/java/source_map.go (hunk 5 above)
Comment (florianGla) [reply pkg/java/source_map.go:40]: Good idea. I used a lambda function to get `fd.Close()` to work correctly.

**Review by jochil — APPROVED** [review]
thanks for all the adjustments

**Pull request merged**